onadata/apps/messaging/viewsets.py

```diff
@@ -31,3 +35,35 @@ class MessagingViewSet(mixins.CreateModelMixin, mixins.ListModelMixin,
     filter_backends = (TargetTypeFilterBackend, TargetIDFilterBackend,
                        UserFilterBackend, DjangoFilterBackend)
     filterset_class = ActionFilterSet
+    pagination_class = StandardPageNumberPagination
+
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+        no_of_records = queryset.count()
+        retrieval_threshold = getattr(
+            settings, "MESSAGE_RETRIEVAL_THRESHOLD", 10000)
+        pagination_keys = [self.paginator.page_query_param,
+                           self.paginator.page_size_query_param]
+        query_param_keys = self.request.query_params
+        should_paginate = any(
+            [k in query_param_keys for k in pagination_keys]) or \
+            no_of_records > retrieval_threshold
+        current_page_size = self.request.query_params.get(
+            self.paginator.page_size_query_param) or retrieval_threshold
+        current_page = self.request.query_params.get(
+            self.paginator.page_query_param) or 1
+        headers = generate_pagination_headers(
+            self.request, no_of_records, current_page_size, current_page
+        )
+
+        if should_paginate and \
+                "page_size" not in self.request.query_params.keys():
+            self.paginator.page_size = retrieval_threshold
+
+        if should_paginate:
+            page = self.paginate_queryset(queryset)
+            serializer = self.get_serializer(page, many=True)
+        else:
+            serializer = self.get_serializer(queryset, many=True)
```

Comment:
Does this return everything if pagination params are not provided?

Reply:
It'll return everything if the threshold is not passed i.e if the threshold is 10,000 records; It'll return everything up to 10,000 records and provide pagination headers for easier navigation